# Flight Delay Analysis 1
This notebook contains the analysis of the "On-Time : Reporting Carrier On-Time Performance" dataset, for January 2015.
In terms of technology, we will use pandas for the data manipulation and analysis.

## 1) Data Loading.

In [15]:
import pandas as pd
import numpy as np

We will import the os module, in order to get the directory of this notebook. I use it, to be totally sure of the data path to insert.

In [16]:
import os
cwd = os.getcwd()
cwd

'C:\\Users\\GerardEspejo\\Desktop\\TFM\\TFM'

The dataset is loaded from a local folder "Data", which contains all the used datasets.

In [24]:
df_original = pd.read_csv('C:\\Users\\GerardEspejo\\Desktop\\TFM\\Data\\Jan_2015.csv')
df_original.head()

C:\Users\GerardEspejo\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (77,84,85) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
0  2015        1      1             2            5  2015-01-02   
1  2015        1      1             3            6  2015-01-03   
2  2015        1      1             4            7  2015-01-04   
3  2015        1      1             5            1  2015-01-05   
4  2015        1      1             6            2  2015-01-06   

  OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID OP_CARRIER TAIL_NUM      ...       \
0                NK                  20416         NK   N521NK      ...        
1                NK                  20416         NK   N512NK      ...        
2                NK                  20416         NK   N528NK      ...        
3                NK                  20416         NK   N523NK      ...        
4                NK                  20416         NK   N534NK      ...        

   DIV4_TAIL_NUM  DIV5_AIRPORT  DIV5_AIRPORT_ID  DIV5_AIRPORT_SEQ_ID  \
0            NaN           NaN              NaN                  NaN   
1            NaN           NaN              NaN                  NaN   
2            NaN           NaN              NaN                  NaN   
3            NaN           NaN              NaN                  NaN   
4            NaN           NaN              NaN                  NaN   

  DIV5_WHEELS_ON DIV5_TOTAL_GTIME DIV5_LONGEST_GTIME  DIV5_WHEELS_OFF  \
0            NaN              NaN                NaN              NaN   
1            NaN              NaN                NaN              NaN   
2            NaN              NaN                NaN              NaN   
3            NaN              NaN                NaN              NaN   
4            NaN              NaN                NaN              NaN   

  DIV5_TAIL_NUM  Unnamed: 109  
0           NaN           NaN  
1           NaN           NaN  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN  

[5 rows x 110 columns]

## 2) Checking: content and and data distribution; from the original dataset.

In [26]:
df_original.shape

(469968, 110)

In [27]:
df_original.columns

Index(['YEAR', 'QUARTER', 'MONTH', 'DAY_OF_MONTH', 'DAY_OF_WEEK', 'FL_DATE',
       'OP_UNIQUE_CARRIER', 'OP_CARRIER_AIRLINE_ID', 'OP_CARRIER', 'TAIL_NUM',
       ...
       'DIV4_TAIL_NUM', 'DIV5_AIRPORT', 'DIV5_AIRPORT_ID',
       'DIV5_AIRPORT_SEQ_ID', 'DIV5_WHEELS_ON', 'DIV5_TOTAL_GTIME',
       'DIV5_LONGEST_GTIME', 'DIV5_WHEELS_OFF', 'DIV5_TAIL_NUM',
       'Unnamed: 109'],
      dtype='object', length=110)

It appears that with the full dataset, we have 110 columns. In order to have a more flexible working environment, and because I won't be using all the data available, I will do a pre-filter of the columns needed once generating the dataset. This will help me reduce time and efforts. Therefore, I am creating a more specific dataset to work with.

The new dataset won't have information (columns) related to: Gate Return Information at Origin Airport and Diverted Airport Information. Because we are focusing this analysis in the Delays, I won't be using flight divertions informations, for now.

## 3) Loading the pre-filtered version of the orginal Dataset

In [28]:
df = pd.read_csv('C:\\Users\\GerardEspejo\\Desktop\\TFM\\Data\\Jan_2015_v2.csv')
df.head()

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
0  2015        1      1            22            4  2015-01-22   
1  2015        1      1            22            4  2015-01-22   
2  2015        1      1            22            4  2015-01-22   
3  2015        1      1            22            4  2015-01-22   
4  2015        1      1            22            4  2015-01-22   

  OP_UNIQUE_CARRIER OP_CARRIER TAIL_NUM  OP_CARRIER_FL_NUM     ...      \
0                DL         DL   N969DL               1485     ...       
1                DL         DL   N912DL               1486     ...       
2                DL         DL   N359NW               1487     ...       
3                DL         DL   N957AT               1488     ...       
4                DL         DL   N985DL               1489     ...       

   AIR_TIME  FLIGHTS DISTANCE DISTANCE_GROUP CARRIER_DELAY WEATHER_DELAY  \
0     134.0      1.0    950.0              4           NaN           NaN   
1      90.0      1.0    762.0              4           NaN           NaN   
2     240.0      1.0   1956.0              8           NaN           NaN   
3      29.0      1.0    143.0              1           NaN           NaN   
4     123.0      1.0    689.0              3           NaN           NaN   

   NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY Unnamed: 50  
0        NaN             NaN                  NaN         NaN  
1        NaN             NaN                  NaN         NaN  
2        NaN             NaN                  NaN         NaN  
3        NaN             NaN                  NaN         NaN  
4        NaN             NaN                  NaN         NaN  

[5 rows x 51 columns]

In [29]:
df.shape

(469968, 51)

In [30]:
df.columns

Index(['YEAR', 'QUARTER', 'MONTH', 'DAY_OF_MONTH', 'DAY_OF_WEEK', 'FL_DATE',
       'OP_UNIQUE_CARRIER', 'OP_CARRIER', 'TAIL_NUM', 'OP_CARRIER_FL_NUM',
       'ORIGIN_AIRPORT_ID', 'ORIGIN_CITY_MARKET_ID', 'ORIGIN',
       'ORIGIN_CITY_NAME', 'ORIGIN_STATE_ABR', 'ORIGIN_STATE_NM', 'ORIGIN_WAC',
       'DEST_AIRPORT_ID', 'DEST_CITY_MARKET_ID', 'DEST', 'DEST_CITY_NAME',
       'DEST_STATE_ABR', 'DEST_STATE_NM', 'DEST_WAC', 'CRS_DEP_TIME',
       'DEP_TIME', 'DEP_DELAY', 'DEP_DELAY_NEW', 'DEP_DEL15',
       'DEP_DELAY_GROUP', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY',
       'ARR_DELAY_NEW', 'ARR_DEL15', 'ARR_DELAY_GROUP', 'CANCELLED',
       'CANCELLATION_CODE', 'DIVERTED', 'CRS_ELAPSED_TIME',
       'ACTUAL_ELAPSED_TIME', 'AIR_TIME', 'FLIGHTS', 'DISTANCE',
       'DISTANCE_GROUP', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY',
       'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY', 'Unnamed: 50'],
      dtype='object')

It appears we have reduced the number of columns by 59. Looking at the columns we now have a more easy to use dataset.

In [31]:
df.describe()

YEAR   QUARTER     MONTH   DAY_OF_MONTH    DAY_OF_WEEK  \
count  469968.0  469968.0  469968.0  469968.000000  469968.000000   
mean     2015.0       1.0       1.0      15.853001       4.025559   
std         0.0       0.0       0.0       8.952803       1.933772   
min      2015.0       1.0       1.0       1.000000       1.000000   
25%      2015.0       1.0       1.0       8.000000       2.000000   
50%      2015.0       1.0       1.0      16.000000       4.000000   
75%      2015.0       1.0       1.0      24.000000       6.000000   
max      2015.0       1.0       1.0      31.000000       7.000000   

       OP_CARRIER_FL_NUM  ORIGIN_AIRPORT_ID  ORIGIN_CITY_MARKET_ID  \
count      469968.000000      469968.000000          469968.000000   
mean         2266.351901       12669.404240           31714.181421   
std          1804.269617        1516.976267            1284.310179   
min             1.000000       10135.000000           30070.000000   
25%           760.000000       11292.000000           30627.000000   
50%          1735.000000       12889.000000           31453.000000   
75%          3488.000000       13930.000000           32467.000000   
max          9793.000000       16218.000000           35991.000000   

          ORIGIN_WAC  DEST_AIRPORT_ID     ...            AIR_TIME   FLIGHTS  \
count  469968.000000    469968.000000     ...       457013.000000  469968.0   
mean       55.632303     12669.437149     ...          112.544096       1.0   
std        26.491024      1516.825219     ...           71.682848       0.0   
min         1.000000     10135.000000     ...            8.000000       1.0   
25%        34.000000     11292.000000     ...           60.000000       1.0   
50%        52.000000     12889.000000     ...           94.000000       1.0   
75%        82.000000     13930.000000     ...          144.000000       1.0   
max        93.000000     16218.000000     ...          676.000000       1.0   

            DISTANCE  DISTANCE_GROUP  CARRIER_DELAY  WEATHER_DELAY  \
count  469968.000000   469968.000000   95951.000000   95951.000000   
mean      803.261279        3.684057      17.802368       2.741889   
std       596.249383        2.345556      45.334536      18.510443   
min        31.000000        1.000000       0.000000       0.000000   
25%       366.000000        2.000000       0.000000       0.000000   
50%       641.000000        3.000000       2.000000       0.000000   
75%      1046.000000        5.000000      18.000000       0.000000   
max      4983.000000       11.000000    1971.000000     938.000000   

          NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  Unnamed: 50  
count  95951.000000    95951.000000         95951.000000          0.0  
mean      13.319872        0.069827            22.760211          NaN  
std       24.723111        2.068116            40.939751          NaN  
min        0.000000        0.000000             0.000000          NaN  
25%        0.000000        0.000000             0.000000          NaN  
50%        4.000000        0.000000             4.000000          NaN  
75%       18.000000        0.000000            29.000000          NaN  
max      830.000000      241.000000           948.000000          NaN  

[8 rows x 38 columns]

In [26]:
df_flights_list = df[['YEAR', 'FL_DATE', 'ORIGIN_AIRPORT_ID', 'ORIGIN', 'DEST_AIRPORT_ID', 'DEST', 'DEP_DELAY', 'ARR_DELAY', 'CANCELLED']]
df_flights_list.head(10)

YEAR     FL_DATE  ORIGIN_AIRPORT_ID ORIGIN  DEST_AIRPORT_ID DEST  \
0  2018  2018-01-27              11697    FLL            12266  IAH   
1  2018  2018-01-27              14747    SEA            14771  SFO   
2  2018  2018-01-27              11278    DCA            12266  IAH   
3  2018  2018-01-27              12892    LAX            13930  ORD   
4  2018  2018-01-27              12451    JAX            11618  EWR   
5  2018  2018-01-27              12266    IAH            14107  PHX   
6  2018  2018-01-27              11618    EWR            12173  HNL   
7  2018  2018-01-27              12173    HNL            11618  EWR   
8  2018  2018-01-27              12889    LAS            14771  SFO   
9  2018  2018-01-27              12264    IAD            15304  TPA   

   DEP_DELAY  ARR_DELAY  CANCELLED  
0      -13.0      -12.0        0.0  
1       -4.0      -18.0        0.0  
2       -2.0        1.0        0.0  
3       -9.0       -8.0        0.0  
4      -14.0      -24.0        0.0  
5       -7.0      -19.0        0.0  
6       27.0       19.0        0.0  
7        8.0      -23.0        0.0  
8       -5.0      -22.0        0.0  
9       -7.0      -18.0        0.0

In [27]:
max_delay_mins = df_flights_list[['DEP_DELAY']].max()
max_delay_hours = max_delay_mins/60
max_delay_hours

DEP_DELAY    33.45
dtype: float64

In [28]:
df_flights_list.sort_values(by=['DEP_DELAY'])

YEAR     FL_DATE  ORIGIN_AIRPORT_ID ORIGIN  DEST_AIRPORT_ID DEST  \
56328   2018  2018-01-23              15991    YAK            12523  JNU   
301885  2018  2018-01-03              12889    LAS            14570  RNO   
60266   2018  2018-01-31              15991    YAK            12523  JNU   
303223  2018  2018-01-03              14570    RNO            12889  LAS   
242437  2018  2018-01-11              11433    DTW            11003  CID   
223072  2018  2018-01-05              12519    JMS            11447  DVL   
52411   2018  2018-01-15              15991    YAK            12523  JNU   
250467  2018  2018-01-26              14150    PLN            10333  APN   
51441   2018  2018-01-13              15991    YAK            10926  CDV   
220862  2018  2018-01-20              13241    MEI            14109  PIB   
51900   2018  2018-01-14              10926    CDV            10299  ANC   
59775   2018  2018-01-30              15991    YAK            12523  JNU   
50452   2018  2018-01-11              15991    YAK            12523  JNU   
243875  2018  2018-01-11              14150    PLN            10333  APN   
219874  2018  2018-01-19              12519    JMS            11447  DVL   
53897   2018  2018-01-18              15991    YAK            12523  JNU   
48940   2018  2018-01-08              12819    KTN            14747  SEA   
53878   2018  2018-01-18              15991    YAK            10926  CDV   
53389   2018  2018-01-17              14256    PSG            15841  WRG   
53379   2018  2018-01-17              10926    CDV            10299  ANC   
55344   2018  2018-01-21              10926    CDV            15991  YAK   
232075  2018  2018-01-21              14150    PLN            10333  APN   
55345   2018  2018-01-21              15991    YAK            12523  JNU   
59277   2018  2018-01-29              15991    YAK            12523  JNU   
57821   2018  2018-01-26              14256    PSG            15841  WRG   
215610  2018  2018-01-17              15454    UIN            10967  CGI   
52905   2018  2018-01-16              15991    YAK            12523  JNU   
57832   2018  2018-01-26              10926    CDV            15991  YAK   
56298   2018  2018-01-23              14709    SCC            10754  BRW   
58769   2018  2018-01-28              15991    YAK            10926  CDV   
...      ...         ...                ...    ...              ...  ...   
560967  2018  2018-01-18              15304    TPA            10397  ATL   
561023  2018  2018-01-18              10529    BDL            10397  ATL   
561058  2018  2018-01-18              11057    CLT            10397  ATL   
561060  2018  2018-01-18              14492    RDU            12892  LAX   
561063  2018  2018-01-18              10599    BHM            10397  ATL   
561073  2018  2018-01-18              14869    SLC            10397  ATL   
561122  2018  2018-01-18              10397    ATL            11193  CVG   
561140  2018  2018-01-18              10397    ATL            12264  IAD   
561259  2018  2018-01-18              14321    PWM            10397  ATL   
561275  2018  2018-01-18              14307    PVD            10397  ATL   
561314  2018  2018-01-18              10397    ATL            14492  RDU   
561315  2018  2018-01-18              14492    RDU            10397  ATL   
561316  2018  2018-01-18              10397    ATL            14122  PIT   
561317  2018  2018-01-18              14122    PIT            10397  ATL   
562487  2018  2018-01-08              15304    TPA            12478  JFK   
563132  2018  2018-01-08              14107    PHX            12478  JFK   
563309  2018  2018-01-08              13204    MCO            12478  JFK   
563481  2018  2018-01-08              14635    RSW            12478  JFK   
563749  2018  2018-01-08              11292    DEN            12478  JFK   
563807  2018  2018-01-08              12478    JFK            10721  BOS   
564067  2018  2018-01-08              12478    JFK      

In [33]:
#Check Dataframe dimensions: number of rows
len(df)

570118

In [35]:
#NaN Treatment
df_clean = df.dropna()
df_clean.head()

Empty DataFrame
Columns: [YEAR, QUARTER, MONTH, DAY_OF_MONTH, DAY_OF_WEEK, FL_DATE, OP_UNIQUE_CARRIER, OP_CARRIER, TAIL_NUM, OP_CARRIER_FL_NUM, ORIGIN_AIRPORT_ID, ORIGIN_CITY_MARKET_ID, ORIGIN, ORIGIN_STATE_ABR, ORIGIN_STATE_NM, ORIGIN_WAC, DEST_AIRPORT_ID, DEST_CITY_MARKET_ID, DEST, DEST_STATE_ABR, DEST_STATE_NM, DEST_WAC, CRS_DEP_TIME, DEP_TIME, DEP_DELAY, DEP_DELAY_NEW, DEP_DEL15, DEP_DELAY_GROUP, CRS_ARR_TIME, ARR_TIME, ARR_DELAY, ARR_DELAY_NEW, ARR_DEL15, ARR_DELAY_GROUP, CANCELLED, CANCELLATION_CODE, DIVERTED, CRS_ELAPSED_TIME, ACTUAL_ELAPSED_TIME, AIR_TIME, FLIGHTS, DISTANCE, DISTANCE_GROUP, CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY, Unnamed: 48]
Index: []

[0 rows x 49 columns]

In [37]:
df_flights_list_clean = df_flights_list.dropna()
df_flights_list_clean.head()

YEAR     FL_DATE  ORIGIN_AIRPORT_ID ORIGIN  DEST_AIRPORT_ID DEST  \
0  2018  2018-01-27              11697    FLL            12266  IAH   
1  2018  2018-01-27              14747    SEA            14771  SFO   
2  2018  2018-01-27              11278    DCA            12266  IAH   
3  2018  2018-01-27              12892    LAX            13930  ORD   
4  2018  2018-01-27              12451    JAX            11618  EWR   

   DEP_DELAY  ARR_DELAY  CANCELLED  
0      -13.0      -12.0        0.0  
1       -4.0      -18.0        0.0  
2       -2.0        1.0        0.0  
3       -9.0       -8.0        0.0  
4      -14.0      -24.0        0.0